alternator/ttl.cc

```diff
@@ -769,13 +769,15 @@ future<> expiration_service::run() {
         // in the next iteration by reducing the scanner's scheduling-group
         // share (if using a separate scheduling group), or introduce
         // finer-grain sleeps into the scanning code.
-        std::chrono::seconds scan_duration(std::chrono::duration_cast<std::chrono::seconds>(lowres_clock::now() - start));
-        std::chrono::seconds period(_db.get_config().alternator_ttl_period_in_seconds());
+        std::chrono::milliseconds scan_duration(std::chrono::duration_cast<std::chrono::milliseconds>(lowres_clock::now() - start));
+        std::chrono::milliseconds period(long(_db.get_config().alternator_ttl_period_in_seconds() * 1000));
         if (scan_duration < period) {
             try {
-                tlogger.info("sleeping {} seconds until next period", (period - scan_duration).count());
+                tlogger.info("sleeping {} seconds until next period", (period - scan_duration).count()/1000.0);
                 co_await seastar::sleep_abortable(period - scan_duration, _abort_source);
             } catch(seastar::sleep_aborted&) {}
+        } else {
+                tlogger.warn("scan took {} seconds, longer than period - not sleeping", scan_duration.count()/1000.0);
```

Comment:
Not relate to the implementation, if that happens, it means that either the duration is too short or the scan is too slow.
First, will that mean you'll see the warning over and over again in the log?
Second, can the user do something? If so, we should think about how the user will now

Reply:
I agree, it would be nice to recommend a solution to the user. Can be a trivial follow-up though, so I'll continue queueing this series

Reply:
This could be a monitoring/advisor solution, so maybe we should open an issue over there, the question was if there was something that can be done and if the logs are going to be flood with the same warning

Reply:
Right now the only thing you can do is to except the fate that the scan cannot be done as frequently as you had hoped (and the reason why I warn about this is that you should know your chosen frequency parameter was not obeyed). In the future, we should increase the scanning thread's priority to make it run faster (at the expense of the workload), but we don't have that yet.

You're right that this will "spam" the log once per period (actually, N times per period for N shards). Since the period is normally fairly large (not 0.5 seconds as in our tests), I think this is acceptable. Moreover, it demonstrates an ongoing problem so maybe it's not that bad to get a periodic reminder about it.

Reply:
I mean, as I user, can you change the duration value? Should I?
We can propogate (at least some times we can) a warning through Loki and make an alert out of it.
Alternativerly, we can add a metric and create an alert out of it